lumicks/pylake/population/detail/hmm.py

```diff
@@ -0,0 +1,298 @@
+    """
+    # for here, simpler if states as columns, observations as rows
+    B = np.exp(model.state_log_likelihood(data)).T
+    T = data.size
+    K = model.K
+
+    alpha = np.zeros((T, K))
+    beta = np.ones((T, K))
+    c = np.zeros(T)
+
+    # forward loop, initialize
+    alpha[0] = model.pi * B[0]
+    c[0] = np.sum(alpha[0])
+    alpha[0] = alpha[0] / c[0]
+
+    # forward loop, induction
+    for t in range(1, T):
+        tmp = col(alpha[t - 1]) * model.A
+        alpha[t] = np.sum(tmp, axis=0) * B[t]
+        c[t] = np.sum(alpha[t])
+        alpha[t] = alpha[t] / c[t]
+
+    # backward loop
+    for t in range(1, T):
+        beta[-(t + 1)] = np.sum(model.A * row(B[-t]) * row(beta[-t]), axis=1) / c[-t]
+
+    return alpha, beta, c, B
+
+
+def calculate_temporary_variables(model, alpha, beta, c, B):
+    """Calculate temporary variables for model update and log likelihood of current model.
+
+    Parameters
+    ----------
+    model : ClassicHmm
+        Current model parameters with K states
+    alpha : np.ndarray
+        probability of the sequence y[0]...y[t] and being in state i at time t; shape [T, K]
+    beta : np.ndarray
+        probability of the ending partial sequence y[t+1]...y[T] given starting state i at time t;
+        shape [T, K]
+    c : np.ndarray
+        scaling factors (to avoid underflow); shape [T, ]
+    B : np.ndarray
+        state likelihoods; shape[T, K]
+
+    Returns
+    -------
+    gamma : np.ndarray
+        Probability of being in state i at time t; shape [T, K]
+    xi : np.ndarray
+        Probability of being in state i at time t and transitioning to state j at time t + 1;
+        shape [T, K, K]
+    log_likelihood : float
+    """
+    gamma = alpha * beta  # Eq 27
+
+    # Eq 37
+    xi = (
+        alpha[:-1][:, :, np.newaxis]
+        * model.A
+        * B[1:][:, np.newaxis, :]
+        * beta[1:][:, np.newaxis, :]
+    ) / c[1:][:, np.newaxis, np.newaxis]
+
+    # Eq 103
+    log_likelihood = np.sum(np.log(c))
+
+    return gamma, xi, log_likelihood
+
+
+def viterbi(data, model):
+    """The Viterbi algorithm to assign most probable hidden state path based on observation
+    sequence.
+
+    Details in Equations 29-35.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Observation data, shape [T, ]
+    model : ClassicHmm
+        Model parameters with K states
+
+    Returns
+    -------
+    statepath : np.ndarray
+        Hidden state path, shape [T, ]
+    """
+    T = data.size
+    B = model.state_log_likelihood(data).T
+    psi = np.zeros(B.shape, dtype=int)
+    statepath = np.zeros(T, dtype=int)
+
+    # initialization
+    # for models trained on a single trace you can end up with zeros
+    # if states are well separated or observations have very low noise
+    with np.errstate(divide="ignore"):
+        delta = col(np.log(model.pi) + B[0])
+    # recursion
+    for t in range(1, T):
+        R = delta + np.log(model.A)
```

Comment:
Not so likely for real data, but if a transition is never observed (and this element becomes zero), you could get a runtime warning taking the log here as well.

Reply:
actually it's a real case when two states aren't connected to each other, you get a `0` for that element. added handling for this and a test

Reply:
Excellent